### ISSUE TITLE
Audio Playback (atleast for Android 10)

### ISSUE THREAD

**Issue by firefinchdev** [Genymobile/scrcpy#1407]
Continuing on grounds of #14, Android 10 now has an official API for playback capture.
https://developer.android.com/guide/topics/media/playback-capture

This API is being used by AZ Screen Recorder (for Android 10) and works perfectly.

We can utilize it for playing audio with `scrcpy`.

I am willing to code the android app part for that. Would need assistance for the `scrcpy` part because I am not familiar with the project.

**Comment by rom1v** [comment]
> I am willing to code the android app part for that.

Could find out what's the problem with this PoC: https://github.com/rom1v/qaudio?
See: https://github.com/Genymobile/scrcpy/issues/14#issuecomment-609031452.

**Comment by firefinchdev** [comment]
Yes, I resolved it. Working now. https://github.com/firefinchdev/qaudio

**Comment by firefinchdev** [comment]
It was just that foreground service didn't start due to invalid id, and that AudioRecord was not built properly.

**Comment by rom1v** [comment]
Oh, awesome :tada:

> It was just that foreground service didn't start due to invalid id

This had been reported, but I still had the error when I tested: https://github.com/Genymobile/scrcpy/issues/14#issuecomment-620579761

> AudioRecord was not built properly

Hmm, however, I just tested your branch on my device, it does not work:

```
I/qaudio: START COMMAND
V/AudioPolicy: Null context, checking permission via ActivityManager
V/android.media.AudioRecord: Will record from REMOTE_SUBMIX at full fixed volume
E/IAudioFlinger: createRecord returned error -1
E/AudioRecord: createRecord_l(6): AudioFlinger could not create record track, status: -1
E/AudioRecord-JNI: Error creating AudioRecord instance: initialization check failed with status -1.
E/android.media.AudioRecord: Error code -20 when initializing native AudioRecord object.
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.rom1v.qaudio, PID: 6670
    java.lang.RuntimeException: Unable to start service com.rom1v.qaudio.RecordService@68cf21a with Intent { act=com.rom1v.qaudio.RECORD cmp=com.rom1v.qaudio/.RecordService (has extras) }: java.lang.IllegalStateException: startRecording() called on an uninitialized AudioRecord.
        at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4235)
        at android.app.ActivityThread.access$2200(ActivityThread.java:231)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1997)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7695)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
     Caused by: java.lang.IllegalStateException: startRecording() called on an uninitialized AudioRecord.
        at android.media.AudioRecord.startRecording(AudioRecord.java:1116)
        at com.rom1v.qaudio.RecordService.startRecording(RecordService.java:83)
        at com.rom1v.qaudio.RecordService.onStartCommand(RecordService.java:62)
        at android.app.ActivityThread.handleServiceArgs(ActivityThread.java:4217)
        at android.app.ActivityThread.access$2200(ActivityThread.java:231) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1997) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7695) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950) 
```

**Comment by firefinchdev** [comment]
Well, you need to manually grant the Microphone permission to the app. Actually, I must have explicitly asked the user for the permission, but since its a PoC, I did not do it.

**Comment by rom1v** [comment]
Oh, ok, so a request for permission was missing. Great! Thank you for your help. :+1:

That probably unblocks the situation :)

I will try to work on it when I have some time (which is… not now…).

**Comment by rom1v** [comment]
- https://blog.rom1v.com/2020/06/audio-forwarding-on-android-10/
 - https://github.com/rom1v/sndcpy
